paddle/framework/operator.cc

```diff
@@ -411,7 +411,34 @@ void OperatorWithKernel::Run(const Scope& scope,
                  expected_kernel_key);
   }
 
-  kernel_iter->second->Compute(ctx);
+  if (actual_kernel_key == expected_kernel_key) {
+    kernel_iter->second->Compute(ctx);
+  } else {
+    Scope& op_scope = scope.NewScope();
+    auto input_vars = this->InputVars();
+    for (auto var_name : input_vars) {
+      op_scope.Var(var_name);
+    }
+
+    // TODO(qijun) get appropriate DeviceContext from DeviceContext pool
+    platform::DeviceContext* trans_dev_ctx = nullptr;
+
+    // TODO(qijun) get appropriate DataTransformFn from global map
```

Comment:
This is in progress in https://github.com/PaddlePaddle/Paddle/pull/6953/files

Reply:
Yes, I see. I think that the interface of DataTransFormFn should be like this:
```
 using DataTransformFn = std::function<void(
        const Variable& in, Variable* out, platform::DeviceContext* ctx)>;
```

- we should take variable as parameter, since not all data are LOD_TENSOR.
- DeviceContext should be taken as a parameter to provide necessary handles.

Reply:
yes, the interface will be updated~

Reply:
the two variable may be in two different DeviceContext, is one DeviceContext enough?

Reply:
maybe the interface should be like
```cpp
using DataTransformFn = std::function<
    void(
        const KernelTypePair& pair,
        const platform::ExecutionContext& ctx,
        const Tensor& in, 
        Tensor* out
         )>;
```
DataTransformFn should get device contextes it needed according to ExecutionContext and KernelTypePair

Reply:
maybe the interface should be like
```cpp
using DataTransformFn = std::function<
    void(
        const KernelTypePair& pair,
        const platform::ExecutionContext& ctx,
        const Tensor& in, 
        Tensor* out
         )>;
```
DataTransformFn should get device contextes it needed according to ExecutionContext and KernelTypePair

Reply:
Do you mean such case?
```
MKLOp --> CUDAOp
```
We have to get two `DeviceContext` from global DeviceContext pool. But we can not transform mkl data to cuda data directly. We must transform mkl data to cpu data, then transform cpu data to cuda data. So, here we may have to do transformation twice.

So, the interface could be like this:
```
using DataTransformFn = std::function<void(
        const Variable& in, Variable* out, vector<platform::DeviceContext*> ctx)>;
```

Reply:
@jacquesqiao Let's make the interface cleaner, we can get appropriate DeviceContext according to ExecutionContext and KernelTypePair. 
Let's do it before DataTransform.

Reply:
ok, I think prepare a vector of DeviceContext outside is ok~

Reply:
I think to prepare a vector outside have a problem, the Operator::Run() will have to understand how many and what kind of Devicecontextes the transform_fn needs. On the other hand, the transform_fn need to know which DeviceContext in the Vector it needs for one variable. maybe now we should let the transform_fn itself to handle it, this will be easier and clearer.

Reply:
DataTransformFn is general and should work for every operator and can be used in any other case if we want to transform data. 
So, just let Operator::Run does these dirty work, like getting appropriate DeviceContext, justifying if a variable should be transform or not. Anyway, we have to write such codes. 

> the transform_fn need to know which DeviceContext in the Vector it needs for one variable.

The DataTransformFn do not need to know, the caller of DataTransformFn needs to know. DataTransformFn just transform data.
```
using DataTransformFn = std::function<void(
        const Variable& in, Variable* out, vector<platform::DeviceContext*> ctx)>;
```
The caller has to pass correct variables and device context to DataTransFormFn.